### ISSUE TITLE
How to pass predicate hint from recipient side

### ISSUE THREAD

**Issue by desingho** [delta-io/delta-sharing#247]
We have written a python connector as a recipient using load_as_pandas.
Not able to understand how to pass predicate hints to not load whole table.
Found following info on predicate hint but no code example in python.
https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md#request-body
"The recipient’s client authenticates to the sharing server (via a bearer token or other method) and asks to query a specific table. The client can also provide filters on the data (e.g. “country=US”) as a hint to read just a subset of the data."
how do i pass this filters on data from python connector to delta sharing server ?

**Comment by desingho** [comment]
@api.route("/Product")
class Product(Resource):
    def get(self):
        table_url_product ="http://localhost:8080/delta-sharing/shares/deltashare/schemas/mypoc/tables/product/query"
        payload = {
            "predicateHints": ["ProductUUID = 0a1b0bfe47d1d61b43fbb72ea4870554"],
            "limitHint": 0,
            "version": 0
            }
        headers = {
           "Content-Type": "application/json"
        }
        response = requests.post(url=table_url_product, data=payload, headers=headers)
        return response.json()

I have writtent a REST API /Product, which need to read delta table stored in azure blob storage via delta sharing server and I dont want to read whole table, I just want to send some filter through predicateHints.

In delta-sharing-server side i can see below error which is not clear to me.

1056307 [armeria-common-worker-nio-2-2] ERROR io.delta.sharing.server.DeltaSharingServiceExceptionHandler - Unrecognized token 'predicateHints': was expecting ('true', 'false' or 'null')
 at [Source: predicateHints=ProductUUID+%3D+0a1b0bfe47d1d61b43fbb72ea4870554&limitHint=0&version=0; line: 1, column: 15]
com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'predicateHints': was expecting ('true', 'false' or 'null')
 at [Source: predicateHints=ProductUUID+%3D+0a1b0bfe47d1d61b43fbb72ea4870554&limitHint=0&version=0; line: 1, column: 15]

**Comment by zsxwing** [comment]
Looks like you were sending a GET request. It should be a POST request: https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md#read-data-from-a-table

In addition, please be aware that a delta sharing server is not required to implement the filtering logic for `predicateHints` and `limitHint`. 
- Filtering files based on the SQL predicates is BEST EFFORT. The server may return files that don’t satisfy the predicates.
- Applying limitHint is BEST EFFORT. The server may return files containing more rows than the client requests.
You still need to apply the filtering logic in the client side.

**Comment by desingho** [comment]
I am making a post call only through request.post.
Why delta-sharing-server is throwing error for predicateHints ?
Is there any limited set of criteria which we can pass as a predicateHints ?
For us ProductUUID is a column name so I passed that.

**Comment by zsxwing** [comment]
> I am making a post call only through request.post.

Ah, I missed that. You need to use `json=payload` as it's not a simple python dict. In addition, the response is not a json string. It will be multiple json strings. Hence you would need to use `text` to read it and split it to multiple json strings. You should be able to run the following code:

```
import requests

table_url_product = "https://sharing.delta.io/delta-sharing/shares/delta_sharing/schemas/default/tables/nyctaxi_2019_part/query"

payload = {
  "predicateHints": ["yyyyMM='2019-11'"],
  "limitHint": 0
}
headers = {
  "Content-Type": "application/json; charset=utf-8"
}

response = requests.post(url=table_url_product, json=payload, headers=headers)
print(response.text)
```
> Why delta-sharing-server is throwing error for predicateHints ?

You were sending an illegal format when using `data= payload`. `requests.post` didn't serialize it to a proper json string.

> Is there any limited set of criteria which we can pass as a predicateHints ? For us ProductUUID is a column name so I passed that.

It's up to the server to decide which one it wants to support. The current delta sharing server in this repo has implemented https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md#sql-expressions-for-filtering Note: we disable it on https://sharing.delta.io/ as we open it to the world and don't want to take any risks to evaluate user codes.

**Comment by desingho** [comment]
If I want to add support to filter data based on some column, which code/file in delta-sharing-server will need to be updated ?

**Comment by zsxwing** [comment]
> If I want to add support to filter data based on some column, which code/file in delta-sharing-server will need to be updated ?

The code is in https://github.com/delta-io/delta-sharing/blob/main/server/src/main/scala/io/delta/standalone/internal/PartitionFilterUtils.scala What features do you plan to add? Could you provide an example?

**Comment by desingho** [comment]
Hello Zskwing,

My colleague has posted the use case we are trying to implement through delta-sharing in https://github.com/delta-io/delta-sharing/issues/114
"I have a similar use case where I am trying to consume the Delta Tables into a UI application. Spark wont be an ideal choice here as it takes some time for spark initialization and hence would be difficult to get the data in sub-seconds or seconds. I am looking for some option to pass filters while load_as_pandas. I could pass a filter so that only the portion that is drawn on my chart ui is loaded. Can this be achieved?"

So in short we don't want to use load_as_spark, instead we want to use load_as_pandas and pass some filter in load_as_pandas and load only the subset of data from azure blob storage which we will use to plot some chart on UI.

The PartitionFilterUtils.scala is using spark internally, will this file be used even if the flow is from load_as_pandas method ?

**Comment by desingho** [comment]
I corrected my code and now I am getting response from delta-sharing-server for the /query rest call with predicateHints.
But it seems delta-sharing-server is not considering the predicateHints (ProductUUID='abc') I passed in the payload while returning the result.
Do you think if I partition delta tables based on ProductUUID column then delta-sharing-server will consider this predicateHints for sure while filtering the data ?
What should I do to increase the chance or make it 100% that predicateHints will be respected by delta-sharing-server while returning the result ?

**Comment by zsxwing** [comment]
@desingho did you turn this on for the server? https://github.com/delta-io/delta-sharing/blob/ff7dbbbe37622a13e8fe4064803843cf806ec744/server/src/universal/conf/delta-sharing-server.yaml.template#L45

**Comment by desingho** [comment]
Yes, I did turn this on, but still delta-sharing-server is not filtering the data based on predicateHints.

**Comment by zsxwing** [comment]
What's your predicateHints and did you see any error outputted by this line? https://github.com/delta-io/delta-sharing/blob/main/server/src/main/scala/io/delta/standalone/internal/PartitionFilterUtils.scala#L68

**Comment by zsxwing** [comment]
I saw `ProductUUID = 0a1b0bfe47d1d61b43fbb72ea4870554` in your previous comment. This is not a correct expression. You need to use `ProductUUID = "0a1b0bfe47d1d61b43fbb72ea4870554"` if `ProductUUID` is a string column.

**Comment by desingho** [comment]
def get(self):
        table_url_product ="http://localhost:8080/delta-sharing/shares/deltashare/schemas/poc/tables/product/query"
        payload = {
            "predicateHints": ["ProductUUID = '0a1b0bfe47d1d61b43fbb72ea4870554'"],
            "limitHint": 0
            }
        headers = {
           "Content-Type": "application/json;charset=utf-8"
        }
        response = requests.post(url=table_url_product, json=payload, headers=headers)
        print(response.status_code)
        print(response.text)

I have corrected that mistake and I am getting response also from delta-sharing-server, but the problem is delta-sharing-server is not filtering the data based on the ProductUUID passed in predicateHints.

**Comment by desingho** [comment]
> What's your predicateHints and did you see any error outputted by this line? https://github.com/delta-io/delta-sharing/blob/main/server/src/main/scala/io/delta/standalone/internal/PartitionFilterUtils.scala#L68

I dont see any error on delta-sharing-server side. I am using delta-sharing-server 0.6.0

**Comment by zsxwing** [comment]
> print(response.text)

Does the response contain any sensitive data? If not, could you share it?
